### PR TITLE
Add some reference documentation regarding nested fields

### PR DESCRIPTION
**Pull request by roaks3**
I've personally had trouble remembering how many of our magic-modules configs work for nested fields, so trying to document them a bit more here (if nothing else so I don't need to hold it in my head).

Changes
* `required` is not global, it is scoped to the parent
* `default_from_api` needs to be set on each ancestor

Open questions I wasn't sure about (but would like to add if someone knows):
* Typically, `required` and `default_from_api` don't make sense on the same field, but I think it is needed in some cases for nested fields?
* I'm still unclear how `output` behaves for nested fields

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/api/type.rb [hunk 1]
```diff
@@ -43,6 +43,8 @@ module Fields
       # not attempt to read the field from the API response.
       # NOTE - this doesn't work for nested fields
       attr_reader :url_param_only
+
+      # For nested fields, this only applies within the parent.
```
Comment (rileykarson) [inline mmv1/api/type.rb:47]:
```suggestion
      # For nested fields, this only applies within the parent.
      # For example, an optional parent can contain a required child.
```

**Review by rileykarson — COMMENTED** [review]
Immutable and Output may be worth clarifying too. Immutable applies to just the current level, while output (mostly) applies from the current level down.

**Comment by rileykarson** [comment]
> Typically, required and default_from_api don't make sense on the same field, but I think it is needed in some cases for nested fields

It's common to have a `default_from_api` parent with required children. That means "this block of configuration has a default when not set, and when set you need to write out a full definition".

> I'm still unclear how output behaves for nested fields

I replied w/ my previous comment before reading that. It doesn't quite work how you'd expect! IIRC while it does make all the fields below it output-only for the user, the schema and documentation still get filled out as if they were mutable. We should really validate that it gets set on all children (or obviate doing so by automatically carrying it through)

**Comment by roaks3** [comment]
@rileykarson I used your suggestions and added notes to the `output` and `immutable` fields. I chose not to include the point about how `default_from_api` interacts with `required`, but I can add something if we'd like (it more seems like my own lack of understanding, and would not fit well in these docs).

File: mmv1/api/type.rb [hunk 2]
```diff
@@ -36,13 +36,24 @@ module Fields
...
       attr_reader :removed_message
 
-      attr_reader :output # If set value will not be sent to server on sync
-      attr_reader :immutable # If set to true value is used only on creation
+      # If set value will not be sent to server on sync.
+      # For nested fields, this also needs to be set on each descendant (ie. self,
+      # child, etc.).
+      attr_reader :output
+
+      # If set to true value is used only on creation.
```
Comment (rileykarson) [inline mmv1/api/type.rb:44]:
```suggestion
      # If set to true, changes in the field's value require recreating the resource
```

Comment (rileykarson) [reply mmv1/api/type.rb:44]: The current docs (before your change) describe `ignore_read` more than anything else lol

File: mmv1/api/type.rb [hunk 3]
```diff
@@ -121,6 +132,8 @@ module Fields
       # if true, then we get the default value from the Google API if no value
       # is set in the terraform configuration for this field.
       # It translates to setting the field to Computed & Optional in the schema.
+      # For nested fields, this also needs to be set on each ancestor (ie. self,
```
Comment (rileykarson) [inline mmv1/api/type.rb:135]: Ah- importantly this _applies at a single level_. It's both valid to have a non-O+C field with an O+C child (such as a complex default within a block) and to have an O+C field with a non-O+C child (a required subfield when the block is specified)

Comment (rileykarson) [reply mmv1/api/type.rb:135]: You need to set it through the tree to make the whole tree O+C, which you generally don't need to do.

Comment (roaks3) [reply mmv1/api/type.rb]: Ah ok, I think I just had this wrong then. I've updated, let me know if the wording I chose makes sense.

**Review by rileykarson — APPROVED** [review]

**Pull request merged**